[tree2] Make validation policy injectable (#15722)
microsoft/FluidFramework: experimental/dds/tree2/src/codec/README.md

# Codec

This library defines types and constructs for organizing code that works with persisted data.

> Its fundamental building block is a codec--a bit of code capable of transcoding between some encoded and in-memory format.

Reviewer: Its possible to parse this sentence wrong, and since both formats are used in memory, its also not super clear.

```suggestion
Its fundamental building block is a codec--a bit of code capable of transcoding between a stable format used for persistence and a potentially less stable format only used in memory.
```

Typically, for compatibility reasons, aspects of `SharedTree` should work with `ICodecFamily`s (which can do the above for multiple supported formats).

## JSON Validators

Codecs can define draft 6 JSON schema for their encoded data.

> One policy choice is whether or not encoded data should be validated, and if so--how.

Reviewer: I don't really know em dash grammar, so I looked it up (time to learn that I guess) and I think this usage may be incorrect.

We have `and` working as a coordinating conjunction between two clauses. I don't think "if so--how" is a well-formed clause for the roughly semi-colon equivalent use of em-dash nor the parenthetical phrase use. `if` is working as a subordinating conjunction in this clause, relating the clauses "so" and "how", so I believe it would be more correct to separate those clauses with a comma:

```suggestion
One policy choice is whether or not encoded data should be validated, and if so, how.
```

> Since some `SharedTree` users might not want to pay the bundle-size cost to perform this validation in all environments,
> This policy is injectable using `ICodecOptions` by providing a `JsonValidator`.

Reviewer: ```suggestion
Since some `SharedTree` users might not want to pay the bundle-size cost to perform this validation in all environments,
this policy is injectable using `ICodecOptions` by providing a `JsonValidator`.
```
